홈 화면(오늘의 단어 화면) 피그마 디자인에 맞게 변경

easyhooon opened the pull request:
- 홈 화면 피그마 디자인에 맞게 변경
- LazyColumn 각 아이템의 Key 를 data class 의 hashCode 로 지정

easyhooon committed:
홈 화면 피그마 디자인에 맞게 변경

mwy3055 commented on app/src/main/java/hsk/practice/myvoca/ui/screens/quiz/QuizScreen.kt:
```diff
@@ -122,30 +116,13 @@ private fun QuizNotAvailable(
         )
         Spacer(modifier = Modifier.height(16.dp))
-        Button(
+        AddWordButton(
```
이런건 별도의 커밋으로 빼는게 좋습니다.

mwy3055 replied:
`AddWordButton`을 추가하는 커밋을 별도로 빼는 게 좋았을 것 같습니다. 홈 화면 변경과는 다른 변경사항이기 때문입니다.

mwy3055 commented on app/src/main/java/hsk/practice/myvoca/ui/screens/home/HomeScreen.kt:
```diff
@@ -51,8 +56,10 @@ import kotlinx.collections.immutable.toImmutableList
 import java.time.LocalDateTime
 import java.time.ZoneOffset
 
+// TODO 단어 추가 스낵바?
```
이건 언제 하실 건가요?

mwy3055 commented on app/src/main/java/hsk/practice/myvoca/ui/screens/home/HomeScreen.kt:
```diff
@@ -51,8 +56,10 @@ import kotlinx.collections.immutable.toImmutableList
 import java.time.LocalDateTime
 import java.time.ZoneOffset
 
+// TODO 단어 추가 스낵바?
```
이건 언제 하실 건가요?

easyhooon replied:
홈화면에서 스낵바로 해당 로직을 구현하려면 이전 단어 목록과 현재 단어 목록을 조회한 뒤에 변화가 있으면 띄워주는 식으로 해야할 것 같은데, 그럴려면 로직이 상당히 복잡할 것 같아서 혹시 단어 추가 화면에서 단어를 추가한 뒤에 해당 액티비티가 종료될 때 Toast 를 통해 메세지를 띄우는건 어떨까요? Undo 와 같은 기능이 필요하지 않아서 굳이 snackbar 로 구현하지 않아도 될 것 같다는 생각이 들어서요!

easyhooon replied:
아 StartActivityForResult 로도 구현할 수 있긴 하겠네요 주말에 작업해보겠슴니다

mwy3055 replied:
단어를 비교하는 건 말이 안 되고, startActivityForResult로 구현 가능한지 한번 확인해 주세요. 안되면 toast로 해야...

easyhooon replied:
스낵바 작업이 생각해보니 HomeScreen 에 국한된 기능이 아니어서(AllWordScreen 에서도 ~를 추가했습니다) AllWord 작업 branch 에서 한번에 해도 될것 같슴니다. MainActivity 와 AllWordActivity 사이의 startActivityForResult 를 이용한 상호작용이라 이건 
p.s ~를 수정했습니다는 피그마엔 안보이는군요, 경우의 수가 은근 많네요 머리 아프구만...

mwy3055 replied:
> p.s ~를 수정했습니다는 피그마엔 안보이는군요, 경우의 수가 은근 많네요 머리 아프구만...

수정했을 때도 스낵바가 보이게 해주세요. 문구는 기존 형식처럼 `$word를 수정했습니다.` 로 하면 될 것 같아요

easyhooon replied:
이 작업 #155 에서 진행해서 PR 올리겠습니다 헷갈리게 해서 죄송함다

mwy3055 reviewed (changes requested):
이번 PR부터 정확한 테스트를 위해 제 기기(SM-A528N)에서 점검하기로 했습니다. 

리뷰 드립니다. 정확한 디자인은 피그마를 참고해 주세요.

---

1. (실선 네모) 오늘의 단어가 비어있을 때에만 화면 제목이 보입니다. (`HomeScreen.kt` 123번 라인)
2. (점선 네모) 타이틀의 textStyle은 항상 headlineLarge여야 합니다. (`HomeScreen.kt` 150번 라인)
3. (점선 원) 디자인할 때 아이콘을 Autorenew에서 Refresh로 교체했습니다. 앱에서도 바꿔 주세요.
![image](https://github.com/mwy3055/MyVoca/assets/45386920/19524995-f297-4f6c-8fac-f4e28f4c886b)

4. `TodayWordContent`에서 `alpha`와 `elevation`에 애니메이션이 적용되지 않고 있는 것 같습니다. 애니메이션은 그냥 `animate...AsState` 기본값이 적용되면 됩니다.

5. 체크 버튼을 누르면 간혹 단어가 튀는? 현상이 발생합니다. 

https://github.com/mwy3055/MyVoca/assets/45386920/f93cbdad-81fa-401a-aab2-1f5727b1d767

6. `TodayWordEmptyIndicator`에서 텍스트와 버튼이 딱 붙어 있습니다. 디자인 상으로는 간격을 16dp 줘야 합니다.
![image](https://github.com/mwy3055/MyVoca/assets/45386920/86c7acf6-aff8-4ef7-8eda-ab90a36620aa)

7. (디자인 수정사항) `TodayWordHeader`에서 오늘의 단어가 비어 있을 때도 새로고침 아이콘을 보여줍니다. 새로고침 버튼은 단어가 1개 이상 등록돼있을 때 클릭할 수 있습니다. 클릭했을 때의 행동은 오늘의 단어가 비어있지 않을 때와 같습니다. 자세한 건 피그마 파일을 참고해 주세요.

easyhooon commented:
항상 상세한 코드 리뷰 감사드림니다

mwy3055 commented:
갑자기 든 생각인데, 5번 단어가 튀는 현상은 혹시 LazyColumn에서 아이템을 재활용하는 과정과 관련이 있을까요?

easyhooon committed:
코드 리뷰 반영 1

TodayWordHeader 크기 고정, RefreshIcon 변경, TodayWordEmptyIndicator Text, button 사이에 간격 추가

easyhooon commented:
> 갑자기 든 생각인데, 5번 단어가 튀는 현상은 혹시 LazyColumn에서 아이템을 재활용하는 과정과 관련이 있을까요?

뭐 때문이지 저도 알아봐야겠네요 같은 아이템의 체크박스를 계속 체크 했다 안했다 하면 발생하지 않는데, 다른거 체크하고 돌아와서 체크하면 발생하는군요 다른 아이템과 똑같은 크기로 렌더링이 되었다가 그만큼만으로는 전체를 그리지 못하니까 확장되는 듯한 애니메이션이 발생하는 것 같은데...

easyhooon committed:
코드 리뷰 반영 2

TodayWordTitle 에 등록된 단어의 개수가 출력되도록 롤백, 간격 조정, 단어가 존재하지 않을 때 refreshIcon 이 보이지만 비활성화상태로, 단어가 추가되면 활성화되어 새로고침이 가능하도록 구현

easyhooon commented:
1, 3, 6 번을 [코드 리뷰 반영 1](https://github.com/mwy3055/MyVoca/pull/153/commits/09a93cffcf45b5c9ae555717dcaba4784028d101) 에서 수정 했습니다

2, 7번을 [코드 리뷰 반영 2](https://github.com/mwy3055/MyVoca/pull/153/commits/91eef85f954c25a51d8ff5fb5975ced1296f1a4c) 에서 수정 했습니다. 피그마 HomeScreen 에 ${TodayWordCount}개의 단어가 등록되어 있습니다 가 포함되어 있길래 이전 상태로 롤백 했습니다

4번은 피그마 디자인과 맞추기 위해 alpha 의 위치를 옮겨본 것인데, 현재 코드에서 왜 애니메이션이 적용되지 않는지 잘 모르겠습니다. 롤백해도 마찬가지인것 같은 느낌이군요

mwy3055 commented:
제 예상입니다만, 4번은 recomposition 때문에 애니메이션이 씹히는 게 아닐까요? 확인은 안 해봤습니다

mwy3055 commented:
* 1, 3, 6, 7번 확인했습니다.
* 2번은 제가 `headlineLarge`로 수정해 달라고 말씀드렸는데, `titleLarge`로 수정되었습니다. 그런데 지금이 더 보기 좋은 것 같으니 이대로 가겠습니다. 감사해요
* 4, 5번은 고치기 어려울 것 같으면 말씀해 주세요

easyhooon commented:
>  4, 5번은 고치기 어려울 것 같으면 말씀해 주세요

4, 5 번은 같이 좀 봐주시면 감사하겠습니다

mwy3055 commented:
4, 5번 모두 
* State가 새로 emit됨
* recomposition 발생
* 애니메이션이 씹히거나 뷰가 튐

의 로직으로 발생하는 것 같습니다. 애니메이션이 재생되려고 하다가 바로 넘어가는 걸 보니 95% 그런 것 같습니다. 이건 종강 후에 연구해 보겠습니다.

easyhooon committed:
체크 박스 위치 조정

단어와 그 높아가 같도록

easyhooon committed:
TODO 제거

다른 branch 에서 해당 작업을 완료

easyhooon commented:
해당 branch 는 작업 완료인 것 같슴다

mwy3055 reviewed (approved):
ok

디테일은 다음 PR에서 부탁드립니다
